File: crates/sync/src/decision.rs
/// The outcome of two-axis change detection.
///
/// Local and remote changes are detected independently in their own hash spaces:
/// - `local_changed`:  hash(current file) ≠ content_hash (both in local/markdown format)
/// - `remote_changed`: hash(provider.read()) ≠ remote_hash (both in remote/API format)
///
/// The two hash spaces are NOT comparable because lark-cli's markdown↔richtext
/// conversion is not byte-identical on roundtrip.
#[derive(Debug, Clone, PartialEq)]
pub enum SyncDecision {
    /// Neither side changed — nothing to do.
    NoChange,
    /// Only local changed. Safe to push.
    PushLocal,
    /// Only remote changed. Safe to auto-pull.
    PullRemote,
    /// Both sides changed. User must decide.
    BothModified,
    /// No base hash exists — new file that needs initial push.
    NewFile,
}

/// Pure decision function. No I/O, no side effects.
///
/// - `local_changed`:  true if hash(local file) ≠ content_hash in DB
/// - `remote_changed`: true if hash(provider.read()) ≠ remote_hash in DB
/// - `has_base`:       true if content_hash exists in DB (doc has been synced before)
///
/// When `remote_hash` is unknown (NULL in DB), the caller should pass `remote_changed = false`
/// (conservative: assume remote hasn't changed). This matches the old "optimistic push" behavior.
pub fn decide(local_changed: bool, remote_changed: bool, has_base: bool) -> SyncDecision {
    if !has_base {
        return SyncDecision::NewFile;
    }
    match (local_changed, remote_changed) {
        (false, false) => SyncDecision::NoChange,
        (true, false) => SyncDecision::PushLocal,
        (false, true) => SyncDecision::PullRemote,
        (true, true) => SyncDecision::BothModified,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_no_change() {
        assert_eq!(decide(false, false, true), SyncDecision::NoChange);
    }

    #[test]
    fn test_push_local() {
        assert_eq!(decide(true, false, true), SyncDecision::PushLocal);
    }

    #[test]
    fn test_pull_remote() {
        assert_eq!(decide(false, true, true), SyncDecision::PullRemote);
    }

    #[test]
    fn test_both_modified() {
        assert_eq!(decide(true, true, true), SyncDecision::BothModified);
    }

    #[test]
    fn test_new_file() {
        assert_eq!(decide(true, false, false), SyncDecision::NewFile);
        assert_eq!(decide(false, false, false), SyncDecision::NewFile);
        assert_eq!(decide(true, true, false), SyncDecision::NewFile);
    }

    #[test]
    fn test_remote_unknown_treated_as_unchanged() {
        // When remote_hash is NULL, caller passes remote_changed=false
        // This should result in PushLocal (not BothModified)
        assert_eq!(decide(true, false, true), SyncDecision::PushLocal);
    }
}

File: apps/obsidian/src/sync/decision.ts
/**
 * Two-axis sync change detection.
 *
 * Ported from larknotes `crates/sync/src/decision.rs` — the surviving sync
 * engine of the abandoned standalone desktop app. The lesson it encodes:
 *
 * Local and remote changes are detected independently, each with its OWN
 * signal:
 *   - `localChanged`:  hash(current vault file) ≠ `localBaseHash`
 *                      (content hash — the vault file is local, hashing is free)
 *   - `remoteChanged`: wiki node `obj_edit_time` ≠ `remoteEditTime`
 *                      (timestamp from `wiki nodes list` — no content fetch)
 *
 * The two axes intentionally use different signals. They are not comparable
 * to each other and were never meant to be: `lark-cli`'s markdown roundtrip
 * is canonicalising, so a local hash can't predict a remote hash anyway.
 *
 * History: the pre-increment-1 engine conflated both into one `lastSyncedHash`
 * and recorded `hash(localMd)` after a push — so any Lark normalisation made
 * the next sync see a phantom change. Increment 1 split it into two content
 * hashes; increment 3 replaced the *remote* hash with `obj_edit_time`, which
 * lets a sync detect which docs changed without fetching any of them.
 */
export enum SyncDecision {
  /** Neither side changed — nothing to do. */
  NoChange = "no-change",
  /** Only local changed. Safe to push. */
  PushLocal = "push-local",
  /** Only remote changed. Safe to auto-pull. */
  PullRemote = "pull-remote",
  /** Both sides changed. User must decide. */
  BothModified = "both-modified",
  /** No baseline exists — first time we've seen this node. */
  NewFile = "new-file",
}

/**
 * Pure decision function. No I/O, no side effects.
 *
 * - `localChanged`:  true if hash(local file) ≠ `localBaseHash` in state
 * - `remoteChanged`: true if the node's `obj_edit_time` ≠ `remoteEditTime`
 * - `hasBase`:       true if a `FileSyncState` exists for this node
 *
 * When there is no baseline the caller branches on whether a local file
 * already exists (brand-new pull vs. untracked collision); `decide` only
 * reports `NewFile` so that branch is explicit at the call site.
 */
export function decide(
  localChanged: boolean,
  remoteChanged: boolean,
  hasBase: boolean,
): SyncDecision {
  if (!hasBase) return SyncDecision.NewFile;
  if (!localChanged && !remoteChanged) return SyncDecision.NoChange;
  if (localChanged && !remoteChanged) return SyncDecision.PushLocal;
  if (!localChanged && remoteChanged) return SyncDecision.PullRemote;
  return SyncDecision.BothModified;
}
